### PR TITLE
Stubs out support for produceCampaignCache

### LINKED ISSUE #1844: Message Broker add "campaign_create" and "campaign_update"
Message Broker needs the following field values from any new or updated campaign.

The values will be cached by the Message Broker system and used to assemble the daily digest messages that are tailored for each user based on their campaign activity.

Using hook_node_update() a call to `dosomething_mbp_request('produceCampaignCache', $params)` where:

```
$params = array(
  'application_event' => 'create_campaign' OR 'update_campaign',
  'application_id' => 2, (Drupal site)
  'title' => '??',
  'url' => '??',
  'call_to_action_text' => '??',
  'do_it_text' => '??',
  'cover_image_url' => '??',
  'sponsor_image_url' => '??',
  'update_text' => '??',
  'update_date' => '??',
)
```

**Notes**:
- The cover image needs its own image style to meet email image dimension requirements. Investigate if images can be accessed from user email (http) requests off site.

### PR DESCRIPTION
**Pull request by deezone**
Adding in support for campaign creation and update hooks to send requests / details to Message Broker.

Fixes #1844

See:
https://github.com/DoSomething/message_broker_producer/pull/23
https://github.com/DoSomething/message-broker/issues/36

### CODE REVIEW

**Comment by deezone** [comment]
@aaronschachter Please review this pull request. I think it's a complete solution that will not require any additional logic in the dosomething code base due to the use of hook_node_insert() and hook_node_update().

File: lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module [hunk 1]
```diff
@@ -14,6 +14,44 @@
  *  Optional keyed array of data to send externally.
  */
 function dosomething_mbp_request($origin, $params = NULL) {
+
+  $productionType = NULL;
```
Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module]: minor, but let's stick to Drupal coding standards here, no camelCase. `$production_type`

Comment (deezone) [inline lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module]: Ah, good point. My brain keep flipping between the two standards as I jump between the Drupal side and the OO rabbit code.

File: lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module [hunk 2]
```diff
@@ -14,6 +14,44 @@
...
+/**
+ * Build payload for transactional producer requeusts. Any adjustments to the
+ * data before becoming a payload should be done in this function.
+ *
+ * @param string $origin
+ *  Where the message request came from.
+ * @param array $params
+ *  Optional keyed array of data to send externally.
+ */
+function dosomething_mbp_transactional($origin, $params = NULL) {
```
Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module]: This function name is very general, upon first look it suggests it fires a transactional message.  Let's change the name to reflect what you mention in the comments, maybe something like `dosomething_mbp_get_transactional_array()` ?

File: lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module [hunk 3]
```diff
@@ -67,11 +105,95 @@ function dosomething_mbp_request($origin, $params = NULL) {
...
+
+/**
+ * Build payload for campaign cache message requests.
+ *
+ * @param string $origin
+ *  Where the message request came from.
+ * @param array $params
+ *  Optional keyed array of data to send externally.
+ */
+function dosomething_mbp_campaign_cache($origin, $params = NULL) {
```
Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module]: Same here.  This function name could provide a bit more clarity.  Perhaps something like  `get_campaign_cache_array()` ?

File: lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module [hunk 4]
```diff
@@ -67,11 +105,95 @@ function dosomething_mbp_request($origin, $params = NULL) {
...
+  return $payload;
+}
+
+/**
+ * Build payload for campaign cache message requests.
+ *
+ * @param string $origin
+ *  Where the message request came from.
+ * @param array $params
+ *  Optional keyed array of data to send externally.
```
Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module]: Should also include a `@return array`

File: lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module [hunk 5]
```diff
@@ -14,6 +14,44 @@
...
+
+/**
+ * Build payload for transactional producer requeusts. Any adjustments to the
+ * data before becoming a payload should be done in this function.
+ *
+ * @param string $origin
+ *  Where the message request came from.
+ * @param array $params
+ *  Optional keyed array of data to send externally.
+ */
```
Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module:64]: Should also include a `@return array`

File: lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module [hunk 6]
```diff
@@ -67,11 +105,95 @@ function dosomething_mbp_request($origin, $params = NULL) {
...
   }
-  catch (Exception $e){
-    watchdog('dosomething_mbp', $e, array(), WATCHDOG_ERROR);
-    return FALSE;
+
+  return $payload;
+}
+
+/*
+ * Impliments hook_node_insert()
```
Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module]: Spelling.  Coding standards also specify that a period should be at the end of the comment, `e.g. Implements hook_node_insert().`

File: lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module [hunk 7]
```diff
@@ -67,11 +105,95 @@ function dosomething_mbp_request($origin, $params = NULL) {
...
+
+/*
+ * Collects and format values from $node submission and populates $params.
+ *
+ * @param string $action
+ *  Type of mbp requet
+ * @param object $node
+ *  Submitted values for campaign
+ */
+function dosomething_mbp_collect_params($action, $node) {
```
Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module]: Same here, this function is doing more than just collecting params, its doing that but also sending a MBP request (which I wouldn't have guessed by the function name).  Perhaps this is something like `dosomething_mbp_campaign_request($action, $node)` ?  Or `set_mbp_campaign_cache` ?  Maybe we should chat in person.. i could see a similar function being used for other node types as well (mainly SMS Game, for example.. if we ever added transactional emails)

File: lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module [hunk 8]
```diff
@@ -67,11 +105,95 @@ function dosomething_mbp_request($origin, $params = NULL) {
...
+ */
+function dosomething_mbp_node_update($node) {
+  dosomething_mbp_collect_params('update_campaign', $node);
+}
+
+/*
+ * Collects and format values from $node submission and populates $params.
+ *
+ * @param string $action
+ *  Type of mbp requet
```
Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module]: Minor, but spelling.

File: lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module [hunk 9]
```diff
@@ -67,11 +105,95 @@ function dosomething_mbp_request($origin, $params = NULL) {
...
+ *
+ * @param object $node
+ *  Submitted values for campaign
+ */
+function dosomething_mbp_node_insert($node) {
+  dosomething_mbp_collect_params('create_campaign', $node);
+}
+
+/*
+ * Impliments hook_node_update()
```
Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module]: Same.

**Comment by deezone** [comment]
@aaronschachter One last review please.

File: lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module [hunk 10]
```diff
@@ -5,15 +5,62 @@
...
+  }
+
+}
+
+/**
+ * Build payload for transactional producer requests.
+ *
+ * NOTE:
+ * Any adjustments to the data before becoming a payload should be done in
+ * this function.
```
Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module:54]: very nice

File: lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module [hunk 11]
```diff
@@ -65,13 +115,103 @@ function dosomething_mbp_request($origin, $params = NULL) {
...
+
+/*
+ * Collects and format values from $node submission and populates $params.
+ *
+ * @param string $action
+ *   Type of mbp request
+ * @param object $node
+ *   Submitted values for campaign
+ */
+function dosomething_mbp_collect_params($action, $node) {
```
Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module]: We should call this function something like `dosomething_mbp_campaign_request` and do the node type check from within inside the `hook_node_[op]` functions.  The function name `collect_params` doesn't suggest its actually sending a request.

Keeping the campaign specific code into one function will prevent this function from becoming a gigantic switch / conditional for the different node types.

File: lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module [hunk 12]
```diff
@@ -65,13 +115,103 @@ function dosomething_mbp_request($origin, $params = NULL) {
...
+ * Implements hook_node_update().
+ *
+ * Respond to the update of campaign nodes to produce Message Broker entries
+ * in the directCacheExchange.
+ *
+ * @param object $node
+ *  Submitted values for campaign
+ */
+function dosomething_mbp_node_update($node) {
+  dosomething_mbp_collect_params('update_campaign', $node);
```
Comment (aaronschachter) [inline lib/modules/dosomething/dosomething_mbp/dosomething_mbp.module]: Per  comment below:

```
if ($node->type == 'campaign') {
  dosomething_mbp_campaign_request('update_campaign', $node);
}
```

**Comment by deezone** [comment]
@aaronschachter Looks like this pull request has become a little old and nasty with other changes in the code base getting pulled in. I assume it would be better to toss this request and open a fresh one?

**Comment by aaronschachter** [comment]
Hmm yikes.  As long as the commits that are not yours have the same hashes, it should be ok.  I suppose it depends on how much work it would be to create a fresh one.  If not, then that's definitely cleaner...

**Pull request closed without merging**